### PR TITLE
Enhance JobAPI with JobObj and ScriptExecutor

### PR DESCRIPTION
**Pull request by SYangster**
- Add JobObj wrapper for related resources
- Removed object-specific parameters from to()
- Support all filetypes when adding scripts, copy flat to custom/ dir for consistency

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/app_common/executors/script_executor.py [hunk 1]
```diff
@@ -34,62 +29,51 @@
...
-                    self._to_nvflare_converter = KerasModelToNumpyParamsConverter(
-                        [AppConstants.TASK_TRAIN, AppConstants.TASK_SUBMIT_MODEL]
-                    )
+                self._from_nvflare_converter = NumpyToKerasModelParamsConverter(
+                    [AppConstants.TASK_TRAIN, AppConstants.TASK_VALIDATION]
+                )
+                self._to_nvflare_converter = KerasModelToNumpyParamsConverter(
+                    [AppConstants.TASK_TRAIN, AppConstants.TASK_SUBMIT_MODEL]
+                )
         # TODO: support other params_exchange_format
```
Comment (YuanTingHsieh) [inline nvflare/app_common/executors/script_executor.py]: if no self._from_nvflare_converter and no self._to_nvflare_converter, what will happen?

File: nvflare/app_common/executors/script_executor.py [hunk 2]
```diff
@@ -34,62 +29,51 @@
...
+        in-process `launch_external_process=False` uses InProcessClientAPIExecutor (default).
+        ex-process `launch_external_process=True` uses ClientAPILauncherExecutor.
 
         Args:
+            script (str): Script to run. For in-process must be a python script path. For ex-process can be any command supported by python subprocess.
+            script_args (str): Optional arguments for script.
+            launch_external_process (bool): Whether to launch the script in external process. Defaults to False.
+            params_transfer_type (str): How to transfer the parameters. FULL means the whole model parameters are sent.
+                DIFF means that only the difference is sent. Defaults to TransferType.FULL.
+            params_exchange_format (str): Format to exchange the parameters. Defaults based on detected imports.
```
Comment (YuanTingHsieh) [inline nvflare/app_common/executors/script_executor.py]: I personally prefer to use something more general like `ExecutorJobObj` to pass in either InProcessClientAPIExecutor or LauncherExecutor, if those names are not good, we just rename them, instead of keep wrapping into new classes?

And this exchange format could be useful in other type of executor as well.

File: nvflare/fuel/utils/pipe/cell_pipe.py [hunk 3]
```diff
@@ -178,18 +179,22 @@ def __init__(
...
 
+        self.site_name = site_name
+        self.token = token
+        self.root_url = root_url
+        self.secure_mode = secure_mode
+        self.workspace_dir = workspace_dir
+
+        sysvarname_placeholders = ["{" + varname + "}" for varname in dir(SystemVarName)]
+        if any([arg in sysvarname_placeholders for arg in [site_name, token, root_url, secure_mode, workspace_dir]]):
+            return
```
Comment (YuanTingHsieh) [inline nvflare/fuel/utils/pipe/cell_pipe.py:190]: why do we need this part?

File: nvflare/job_config/fed_job.py [hunk 4]
```diff
@@ -349,6 +383,91 @@ def _add_referenced_components(self, base_component, target):
                             # remove already added components from tracked components
                             self._components.pop(base_id)
 
+    def _add_script_executor(self, obj, target, tasks):
+        if obj._launch_external_process:
```
Comment (YuanTingHsieh) [inline nvflare/job_config/fed_job.py:387]: Should we just make this `_add_executor_obj`?
Rename/merge `ScriptExecutor` with `ExecutorJobObj` since both of them are just wrappers?

File: nvflare/job_config/base_app_config.py [hunk 5]
```diff
@@ -58,9 +58,6 @@ def add_ext_script(self, ext_script: str):
         if not (os.path.isabs(ext_script) or os.path.exists(ext_script)):
             raise RuntimeError(f"Could not locate external script: {ext_script}")
 
-        if not ext_script.endswith(".py"):
-            raise RuntimeError(f"External script: {ext_script} must be a '.py' file.")
-
         self.ext_scripts.append(ext_script)
 
     def add_ext_dir(self, ext_dir: str):
```
Comment (YuanTingHsieh) [inline nvflare/job_config/base_app_config.py:63]: what's special about "ext_scripts" or "ext_dir", should they just be "resources"?

File: examples/getting_started/pt/src/cifar10_lightning_ddp_fl.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import torchvision
+import torchvision.transforms as transforms
+from lit_net import LitNet
+from pytorch_lightning import LightningDataModule, Trainer, seed_everything
+from torch.utils.data import DataLoader, random_split
+
+# (1) import nvflare lightning client API
+import nvflare.client.lightning as flare
+
+seed_everything(7)
```
Comment (YuanTingHsieh) [inline examples/getting_started/pt/src/cifar10_lightning_ddp_fl.py:25]: It this file relate to the PR?

File: nvflare/__init__.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 
 # https://github.com/microsoft/pylance-release/issues/856
 
-from nvflare.app_common.executors.script_executor import ScriptExecutor
 from nvflare.app_common.workflows.fedavg import FedAvg
 from nvflare.job_config.fed_job import FedJob, FilterType
+from nvflare.job_config.script_executor import ScriptExecutor
```
Comment (YuanTingHsieh) [inline nvflare/__init__.py:23]: I am not sure if we want them here.
What are the special things we want to do "from nvflare import XXX"

File: nvflare/job_config/script_executor.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                )
+        if tf_ok:
+            if params_exchange_format == ExchangeFormat.NUMPY:
+                self._from_nvflare_converter = NumpyToKerasModelParamsConverter(
+                    [AppConstants.TASK_TRAIN, AppConstants.TASK_VALIDATION]
+                )
+                self._to_nvflare_converter = KerasModelToNumpyParamsConverter(
+                    [AppConstants.TASK_TRAIN, AppConstants.TASK_SUBMIT_MODEL]
+                )
+        # TODO: support other params_exchange_format
```
Comment (YuanTingHsieh) [inline nvflare/job_config/script_executor.py:79]: if no self._from_nvflare_converter and no self._to_nvflare_converter, what will happen?

File: nvflare/job_config/script_executor.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        ex-process `launch_external_process=True` uses ClientAPILauncherExecutor.
+
+        Args:
+            script (str): Script to run. For in-process must be a python script path. For ex-process can be any command supported by python subprocess.
+            script_args (str): Optional arguments for script.
+            launch_external_process (bool): Whether to launch the script in external process. Defaults to False.
+            params_transfer_type (str): How to transfer the parameters. FULL means the whole model parameters are sent.
+                DIFF means that only the difference is sent. Defaults to TransferType.FULL.
+            params_exchange_format (str): Format to exchange the parameters. Defaults based on detected imports.
+        """
```
Comment (YuanTingHsieh) [inline nvflare/job_config/script_executor.py:53]: I personally prefer to use something more general like ExecutorJobObj to pass in either InProcessClientAPIExecutor or LauncherExecutor, if those names are not good, we just rename them, instead of keep wrapping into new classes?

And this exchange format could be useful in other type of executor as well.

File: nvflare/app_opt/pt/params_converter.py [hunk 10]
```diff
@@ -12,15 +12,21 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Dict
+from typing import Dict, List
```
Comment (yanchengnv) [inline nvflare/app_opt/pt/params_converter.py:15]: Is this file related to this PR?

File: nvflare/fuel/utils/pipe/cell_pipe.py (hunk 3 above)
Comment (yanchengnv) [reply nvflare/fuel/utils/pipe/cell_pipe.py:190]: This file doesn't seem relevant to this PR. If you see any changes needed, please open a different PR.

File: nvflare/job_config/base_app_config.py (hunk 5 above)
Comment (yanchengnv) [reply nvflare/job_config/base_app_config.py:63]: I think they are treated differently when copied into custom.

File: nvflare/job_config/fed_job.py [hunk 11]
```diff
@@ -14,22 +14,30 @@
...
-from typing import Any, List, Union
+from typing import Any, List, Optional, Union
 
 from nvflare.apis.executor import Executor
-from nvflare.apis.filter import Filter
+from nvflare.apis.filter import Filter, FilterType
 from nvflare.apis.impl.controller import Controller
 from nvflare.apis.job_def import ALL_SITES, SERVER_SITE_NAME
-from nvflare.app_common.executors.script_executor import ScriptExecutor
+from nvflare.app_common.executors.client_api_launcher_executor import ClientAPILauncherExecutor
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job.py:23]: Why do we need to import those special Executors? If we import any special Executor, we are doing something wrong.

File: nvflare/job_config/fed_job.py [hunk 12]
```diff
@@ -14,22 +14,30 @@
...
+from nvflare.app_common.executors.in_process_client_api_executor import InProcessClientAPIExecutor
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job.py:24]: We need serious rework on this. This long list of imports implies that something is wrong with the design. We cannot possibly take all the details of each special object into this framework level code!

File: nvflare/job_config/fed_job.py [hunk 13]
```diff
@@ -213,33 +250,36 @@ def to(
...
+                if target not in self._deploy_map:
+                    raise ValueError(
+                        f"{target} doesn't have an `Executor`. Deploy one first before adding client-side controllers!"
+                    )
+                self._deploy_map[target].add_component(obj, id)
+            else:
+                if target not in self._deploy_map:
+                    self._deploy_map[target] = ControllerApp(key_metric=self.key_metric)
+                self._deploy_map[target].add_controller(obj, id)
+        elif isinstance(obj, (Executor, ScriptExecutor)):
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job.py:263]: Why do we care about ScriptExecutor here?

File: nvflare/job_config/fed_job.py [hunk 14]
```diff
@@ -349,6 +383,91 @@ def _add_referenced_components(self, base_component, target):
                             # remove already added components from tracked components
                             self._components.pop(base_id)
 
+    def _add_script_executor(self, obj, target, tasks):
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job.py:386]: This logic doesn't belong here! You can define a method like add_to_job() for JobObj class, and let the ScriptExecutorJobObj implement this detail in its own way.

The fed_job module is framework level code. It should be coded based on abstraction. If what we defined for JobObj is not enough, you can expand its abstraction.

File: nvflare/job_config/script_executor.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (yanchengnv) [inline nvflare/job_config/script_executor.py:1]: This file does not belong to nvflare.job_config. It should go to somewhere in nvflare.app_opt.

File: nvflare/job_config/fed_job.py [hunk 16]
```diff
@@ -213,33 +250,36 @@ def to(
...
+                )
             if target not in self._deploy_map:
                 self._deploy_map[target] = ExecutorApp()
+
             if isinstance(obj, ScriptExecutor):
-                external_scripts = [obj._task_script_path]
-                self._deploy_map[target].add_external_scripts(external_scripts)
+                self._add_script_executor(obj, target, tasks)
+            else:
+                self._deploy_map[target].add_executor(obj, tasks=tasks)
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job.py:274]: Shouldn't this be obj.base_obj?

**Review by yanchengnv — CHANGES_REQUESTED** [review]
The fed_job module needs rework. See my comments. Make sure all modules in nvflare.job_config is framework level code and does not contain any special-purpose executor type.

**Comment by SYangster** [comment]
This PR was originally branched from https://github.com/NVIDIA/NVFlare/pull/2751 (combining InProcessClientAPIExecutor or ClientAPILauncherExecutor in ScriptExecutor), but was based on a prior design before adding JobObj.

Rebased to remove those changes from this branch.

File: nvflare/fuel/utils/pipe/cell_pipe.py (hunk 3 above)
Comment (SYangster) [reply nvflare/fuel/utils/pipe/cell_pipe.py]: removed

File: nvflare/job_config/fed_job.py (hunk 12 above)
Comment (SYangster) [reply nvflare/job_config/fed_job.py]: removed from this pr, will refactor fed_job.py to improve layering

File: nvflare/job_config/fed_job.py (hunk 16 above)
Comment (SYangster) [reply nvflare/job_config/fed_job.py]: obj.base_obj is already passed into _to() from to()

File: nvflare/__init__.py [hunk 17]
```diff
@@ -18,7 +18,7 @@
 
 # https://github.com/microsoft/pylance-release/issues/856
 
-from nvflare.app_common.executors.script_executor import ScriptExecutor
 from nvflare.app_common.workflows.fedavg import FedAvg
+from nvflare.app_opt.script_executor import ScriptExecutor
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/__init__.py:22]: ScriptExecutor imports many modules from app_opt. Stuff from app_common shouldn't import things from app_opt.

File: nvflare/job_config/fed_job.py [hunk 18]
```diff
@@ -14,22 +14,24 @@
...
 from nvflare.app_common.widgets.intime_model_selector import IntimeModelSelector
 from nvflare.app_common.widgets.validation_json_generator import ValidationJsonGenerator
+from nvflare.app_opt.script_executor import ScriptExecutor
 from nvflare.fuel.utils.class_utils import get_component_init_parameters
 from nvflare.fuel.utils.import_utils import optional_import
 from nvflare.fuel.utils.validation_utils import check_positive_int
 from nvflare.job_config.fed_app_config import ClientAppConfig, FedAppConfig, ServerAppConfig
 from nvflare.job_config.fed_job_config import FedJobConfig
 
+from .job_object import ExecutorJobObj, FilterJobObj, JobObj
```
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/job_config/fed_job.py:33]: torch should not be imported here.
